components/ai_chat/core/browser/ai_chat_database.cc

```diff
@@ -605,68 +613,106 @@ bool AIChatDatabase::AddConversation(mojom::ConversationPtr conversation,
+    static constexpr char kSelectExistingAssociatedContentIds[] =
+        "SELECT uuid FROM associated_content WHERE conversation_uuid=?";
+    sql::Statement select_existing_ids(GetDB().GetCachedStatement(
+        SQL_FROM_HERE, kSelectExistingAssociatedContentIds));
+    select_existing_ids.BindString(0, conversation_uuid);
+    std::vector<std::string> existing_ids;
+    while (select_existing_ids.Step()) {
+      existing_ids.push_back(select_existing_ids.ColumnString(0));
+    }
+
+    // Store as a set for faster lookup. Note: We don't push directly to the set
+    // as that is O(n**2) with a base::flat_set.
+    existing_ids_set = existing_ids;
+  }
+
+  for (size_t i = 0; i < associated_content.size(); ++i) {
+    auto& content = associated_content[i];
+    auto content_text = contents.empty() ? "" : std::move(contents[i]);
+
+    sql::Statement insert_or_update_statement;
+    bool exists = existing_ids_set.contains(content->uuid);
+    if (exists) {
+      DVLOG(4) << "Updating associated content for conversation "
+               << conversation_uuid << " with id " << content->uuid;
+      static constexpr char kUpdateAssociatedContentQuery[] =
+          "UPDATE associated_content"
+          " SET title = ?,"
+          " url = ?,"
+          " content_type = ?,"
+          " last_contents = ?,"
+          " content_used_percentage = ?,"
+          " is_content_refined = ?"
+          " WHERE uuid=? and conversation_uuid=?";
+      insert_or_update_statement.Assign(
+          GetDB().GetUniqueStatement(kUpdateAssociatedContentQuery));
+    } else {
+      DVLOG(4) << "Inserting associated content for conversation "
+               << conversation_uuid;
+      static constexpr char kInsertAssociatedContentQuery[] =
+          "INSERT INTO associated_content(title, url,"
+          " content_type, last_contents, content_used_percentage,"
+          " is_content_refined, uuid, conversation_uuid)"
+          " VALUES(?, ?, ?, ?, ?, ?, ?, ?) ";
+      insert_or_update_statement.Assign(
+          GetDB().GetUniqueStatement(kInsertAssociatedContentQuery));
+    }
+
+    CHECK(insert_or_update_statement.is_valid());
+    int index = 0;
+    BindAndEncryptOptionalString(insert_or_update_statement, index++,
+                                 content->title);
+    BindAndEncryptOptionalString(insert_or_update_statement, index++,
+                                 content->url.spec());
+    insert_or_update_statement.BindInt(
+        index++, base::to_underlying(content->content_type));
+    BindAndEncryptOptionalString(insert_or_update_statement, index++,
+                                 content_text);
+    insert_or_update_statement.BindInt(index++,
+                                       content->content_used_percentage);
+    insert_or_update_statement.BindBool(index++, content->is_content_refined);
+    insert_or_update_statement.BindString(index++, content->uuid);
+    insert_or_update_statement.BindString(index, conversation_uuid);
```

Comment:
does this have some kind of SQL injection protection? given that content/title is essentially anything the page wants it to be AFAICT

Reply:
This is a prepared statement, so I think this should be fine. @petemill to confirm

Reply:
Yes, these statements are precompiled prepared statements and the AST structure cannot be modified by the values. The `BindX` calls `sqlite3_bind_x` which adds a parameter value to the statement.